keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,821 @@
+causing the node to retain outdated information and leading to inconsistencies or an imbalance in the cluster, thus affecting the optimal scheduling and deployment of workloads. As a side-effect, it is also possible for the workloads
+to be force migrated to a different node, causing a temporary spike in the CPU/Memory utilisation which is undesirable.
+
+With the current state of implementation in the Kubernetes realm, the available workarounds to allow the cluster to be aware of the changes in the cluster's capacity is by 
+restarting the node or at-least restarting the kubelet, which does not have a certain set of best-practices to follow.
+
+However, this approach does carry a few drawbacks such as
+ - Introducing a downtime for the existing/to-be-scheduled workloads on the cluster until the node is available.
+ - Necessity to reconfigure the underlying services post node-reboot.
+ - Managing the associated nuances that a kubelet restart or node reboot carries such as
+   - https://github.com/kubernetes/kubernetes/issues/109595
+   - https://github.com/kubernetes/kubernetes/issues/119645
+   - https://github.com/kubernetes/kubernetes/issues/125579
+   - https://github.com/kubernetes/kubernetes/issues/127793
+
+Hence, it is necessary to handle the updates in the compute capacity in a graceful fashion across the cluster, than adopting to reset the cluster components to achieve the same.
+
+Also, given that the capability to live-resize a node exists in the Linux and Windows kernels, enabling the kubelet to be aware of the underlying changes in the node's compute capacity will mitigate any actions that are required to be made
+by the Kubernetes administrator.
+
+Node resource hot plugging proves advantageous in scenarios such as:
+- Efficiently managing resource demands with a limited number of nodes by increasing the capacity of existing nodes instead of provisioning new ones.
+- The procedure of establishing new nodes is considerably more time-intensive than expanding the capabilities of current nodes.
+- Improved inter-pod network latencies as the inter-node traffic can be reduced if more pods can be hosted on a single node.
+- Easier to manage the cluster with fewer nodes, which brings less overhead on the control-plane
+- Mitigate a few of the existing limitations/issues that are associated with a node/kubelet restart.
+
+Implementing this KEP will empower nodes to recognize and adapt to changes in their compute configurations and allow facilitating the efficient and effective deployment of pod workloads to nodes capable of meeting the required compute demands.
+
+### Goals
+
+* Achieve seamless node capacity expansion through hot plugging resources.
+* Enable the re-initialization of resource managers (CPU manager, memory manager) to accommodate alterations in the node's resource allocation.
+* Recalculating the OOMScoreAdj and swap memory limit for existing pods.
```

Comment:
would those two result in a new CRI API calls? Which ones?

Reply:
We decided to use `UpdateContainerResources` CRI method for updating both OOMScoreAdj and Swap, Initial plan is to serially update the containers across Pods.
Incase of errors to have retry mechanism like SynchPod.

Reply:
Sure definitely, We will update the KEP with more information.